KAFKA-18984: Reset interval.ms By Using kafka-client-metrics.sh

Parkerhiphop opened the pull request:
**Problem**

kafka-client-metrics.sh cannot reset the interval using `--interval=`.

  - `./bin/kafka-client-metrics.sh --alter --bootstrap-server=$BROKERS --name $NAME  --interval=` results in: ERROR joptsimple.OptionArgumentConversionException: Cannot parse argument '' of option interval.

Meanwhile, `--metrics=` and `--match=` can successfully reset their respective configurations.

**Expected behavior**

Allow resetting the interval configuration by passing an empty value (`--interval=`).
- `--interval=6000 `→ Sets the interval to 6000 (validated as a valid integer)
- `--interval=` → Resets (deletes) the interval setting
- `--interval=abc` → Fails with an error about invalid interval value

**Solution**
- Changing the `Interval` type from Integer to String to handle empty values.
  - [jopt-simple](https://github.com/jopt-simple/jopt-simple) cannot parse an empty string from `--interval=` as an Integer. When trying to parse an empty string as an integer, it throws a conversion exception before the code gets a chance to handle it.
- Validate the interval value in the `checkArgs()` like other opts.
- Add the `testAlterResetConfigs` and `testOptionsAlterInvalidInterval` to cover the changes.

Parkerhiphop committed:
KAFKA-18984: Reset interval.ms By Using kafka-client-metrics.sh

chia7712 reviewed (commented):
@Parkerhiphop please add test for this bug fix!

Parkerhiphop committed:
Add test

Parkerhiphop committed:
Merge branch 'upstream-trunk' into KAFKA-18984

Parkerhiphop committed:
Add alter reset configs test

Parkerhiphop committed:
Rm redundant test

Parkerhiphop commented:
> @Parkerhiphop please add test for this bug fix!

@chia7712  Thanks for the reminder.
I’ve added `testAlterResetConfigs` to test that empty options can reset configs, and `testOptionsAlterInvalidInterval` to verify that an invalid `--interval=` throws a valid integer error.

I also moved the interval invalid error to `checkArgs`, as it is more suitable there than in `alterClientMetrics`.

chia7712 commented:
@Parkerhiphop it seems to me this PR does NOT fix the issue actually, and so I add "WIP" to it.

chia7712 commented:
Sorry that I guess the change of type is used to support "xxxx=`, right?

Parkerhiphop commented:
@chia7712 Yes! Sorry for the vague description earlier. I’ve updated the PR description to provide a clearer explanation.

Parkerhiphop commented:
Local testing screenshot:

![image](https://github.com/user-attachments/assets/537dd9b6-0f63-48ad-90d8-0b77d41a6bae)

For valid and invalid string
![CleanShot 2025-03-17 at 11 38 49@2x](https://github.com/user-attachments/assets/155e623c-208d-49a1-b5ac-e4b1fd631f37)

m1a2st commented on tools/src/main/java/org/apache/kafka/tools/ClientMetricsCommand.java:
```diff
@@ -241,7 +241,7 @@ public ClientMetricsCommandOptions(String[] args) {
             intervalOpt = parser.accepts("interval", "The metrics push interval in milliseconds.")
```
Could you add this "reset the interval" usage into document?

chia7712 replied:
we should do that for other configs as well.

Parkerhiphop replied:
@m1a2st @chia7712 Thanks for the reminder.
I've added the `Leave empty to reset the XXX` for them into the document.

m1a2st reviewed (commented):
Thanks for this patch, left a comment

Parkerhiphop committed:
Add reset description for interval, match and metrics

Parkerhiphop committed:
Merge branch 'upstream-trunk' into KAFKA-18984

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/ClientMetricsCommandTest.java:
```diff
@@ -156,6 +166,27 @@ public void testAlterGenerateName() {
+        ClientMetricsCommand.ClientMetricsService service = new ClientMetricsCommand.ClientMetricsService(adminClient);
+
+        AlterConfigsResult result = AdminClientTestUtils.alterConfigsResult(new ConfigResource(ConfigResource.Type.CLIENT_METRICS, clientMetricsName));
+        when(adminClient.incrementalAlterConfigs(any(), any())).thenReturn(result);
```
Could you please capture the input arguments to ensure all of them are delete ops?

Parkerhiphop replied:
@chia7712 I've add an ArgumentCaptor to capture the input arguments to ensure all of them are delete ops.
Thanks for reviewing.

Parkerhiphop committed:
capture input argument and test strictly

Parkerhiphop committed:
Merge remote-tracking branch 'upstream/trunk' into KAFKA-18984

Parkerhiphop commented:
Since the reset behavior has changed, it now resets the configs to their default values instead of leaving them empty as before.

Here’s a screenshot of the reset behavior—interval.ms is reset to its default value (300000).

![CleanShot 2025-03-21 at 01 42 01@2x](https://github.com/user-attachments/assets/2e5a21d0-1165-4e85-b62f-ae7a60852e94)

apoorvmittal10 commented on tools/src/main/java/org/apache/kafka/tools/ClientMetricsCommand.java:
```diff
@@ -367,6 +368,16 @@ public void checkArgs() {
+                        try {
+                            Integer.parseInt(intervalStr);
+                        } catch (NumberFormatException e) {
+                            throw new IllegalArgumentException("Invalid interval value. Must be a valid integer or empty to delete the setting.");
```
Can this be better to specify?
```suggestion
                            throw new IllegalArgumentException("Invalid interval value. Enter an integer, or leave empty to delete.");
```

Parkerhiphop replied:
@apoorvmittal10  Thanks for the suggestion!

Instead of vaguely stating that it should be a valid integer or empty, the new message clearly specifies that it must be an integer or left empty to delete.

Since empty values now reset the interval to its default value (as shown in the previous screenshot), I think we can explicitly say “reset” here. What do you think?
```
IllegalArgumentException("Invalid interval value. Enter an integer, or leave empty to reset.");
```

apoorvmittal10 commented on tools/src/test/java/org/apache/kafka/tools/ClientMetricsCommandTest.java:
```diff
@@ -156,6 +168,41 @@ public void testAlterGenerateName() {
+        ClientMetricsCommand.ClientMetricsService service = new ClientMetricsCommand.ClientMetricsService(adminClient);
+
+        AlterConfigsResult result = AdminClientTestUtils.alterConfigsResult(new ConfigResource(ConfigResource.Type.CLIENT_METRICS, clientMetricsName));
+        @SuppressWarnings("unchecked")
+        final ArgumentCaptor<Map<ConfigResource, Collection<AlterConfigOp>>> configCaptor = ArgumentCaptor.forClass(Map.class);
+        when(adminClient.incrementalAlterConfigs(configCaptor.capture(), any())).thenReturn(result);
```
Query: Can you please help explain regarding what we want to test here?

Parkerhiphop replied:
@apoorvmittal10  This test verifies the behavior of resetting (deleting) client metrics configurations, simulating a scenario where a user wants to reset all configurations by passing empty values:
```
"--metrics", "",
"--interval", "",
"--match", ""
```
The `ArgumentCaptor` is used to inspect the exact configuration changes sent to Kafka, ensuring that the command-line tool correctly translates user input into the appropriate admin client operations.

In `when(adminClient.incrementalAlterConfigs(configCaptor.capture(), any()))`, the `capture()` method captures the argument passed to `incrementalAlterConfigs()` for later verification.

The test performs the following checks:
- Ensures that exactly one `ConfigResource` is being modified.
- Verifies that exactly three operations (one for each config: `metrics`, `interval`, and `match`) are performed.
- Confirms that all operations are of type `DELETE`, meaning the configurations are being removed.
- Validates that the command outputs a success message indicating the configurations were altered.

Let me know if you have any further question!

apoorvmittal10 reviewed (commented):
Thanks for the PR. Looks good, have minor comments.

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/ClientMetricsCommandTest.java:
```diff
@@ -156,6 +168,41 @@ public void testAlterGenerateName() {
+            }
+        });
+        Map<ConfigResource, Collection<AlterConfigOp>> alteredConfigOps = configCaptor.getValue();
+        assertTrue(alteredConfigOps != null, "alteredConfigOps should not be null");
```
`assertNotNull(alteredConfigOps, "alteredConfigOps should not be null");`

chia7712 commented on tools/src/test/java/org/apache/kafka/tools/ClientMetricsCommandTest.java:
```diff
@@ -156,6 +168,41 @@ public void testAlterGenerateName() {
+        
+        Collection<AlterConfigOp> operations = alteredConfigOps.values().iterator().next();
+        assertEquals(3, operations.size(), "Should have exactly 3 operations");
+        for (Collection<AlterConfigOp> ops : alteredConfigOps.values()) {
```
```java
        for (AlterConfigOp op : operations) {
            assertEquals(AlterConfigOp.OpType.DELETE, op.opType(),
                    "Expected DELETE operation for config: " + op.configEntry().name());
        }
```

Parkerhiphop replied:
@chia7712 Thanks for pointing it out, I've updated the test accordingly.

Parkerhiphop committed:
Fix test

Parkerhiphop committed:
Specify exception better

Parkerhiphop committed:
Merge branch 'upstream-trunk' into KAFKA-18984

Parkerhiphop committed:
Fix testOptionsAlterInvalidInterval

m1a2st reviewed (commented):
`ClientMetricsCommand` L123, `.toString()` is redundant, it can be removed

Parkerhiphop committed:
Remove reduntant code

AndrewJSchofield commented on tools/src/main/java/org/apache/kafka/tools/ClientMetricsCommand.java:
```diff
@@ -238,24 +238,25 @@ public ClientMetricsCommandOptions(String[] args) {
                 .ofType(String.class)
                 .withValuesSeparatedBy(',');
-            metricsOpt = parser.accepts("metrics",  "Telemetry metric name prefixes 'm1,m2'.")
+            metricsOpt = parser.accepts("metrics",  "Telemetry metric name prefixes 'm1,m2'." + nl + "Leave empty to reset the metrics prefixes.")
```
I think perhaps this is an unnecessary change. If the list of metric name prefixes is empty, it sets an empty list. This happens to be the default, so it has the effect of "resetting" the prefixes. Personally, I think of it as setting an empty list, which probably doesn't need a change in the usage message.

AndrewJSchofield commented on tools/src/main/java/org/apache/kafka/tools/ClientMetricsCommand.java:
```diff
@@ -238,24 +238,25 @@ public ClientMetricsCommandOptions(String[] args) {
             };
             String matchSelectorNames = Arrays.stream(matchSelectors).map(config -> "\t" + config).collect(Collectors.joining(nl));
-            matchOpt = parser.accepts("match",  "Matching selector 'k1=v1,k2=v2'. The following is a list of valid selector names: " + nl + matchSelectorNames)
+            matchOpt = parser.accepts("match",  "Matching selector 'k1=v1,k2=v2'. The following is a list of valid selector names: " + nl + matchSelectorNames + nl + "Leave empty to reset the match selector.")
```
I think perhaps this is an unnecessary change. If the list of match selectors is empty, it sets an empty list. This happens to be the default, so it has the effect of "resetting" the selectors. Personally, I think of it as setting an empty list, which probably doesn't need a change in the usage message.

Parkerhiphop replied:
Thanks for the clarification!
That makes sense. I’ve reverted the changes accordingly.

AndrewJSchofield reviewed (changes requested):
Thanks for the PR. With the exception of a couple of comments about the usage message, I think this is looking good.

Parkerhiphop committed:
Merge branch 'trunk' into KAFKA-18984

Parkerhiphop committed:
Remove unnecessary changes

AndrewJSchofield reviewed (approved):
lgtm

showuon commented:
@AndrewJSchofield @apoorvmittal10 , do you think we should backport this patch to 3.9.1 release?

chia7712 commented:
> do you think we should backport this patch to 3.9.1 release?

This is a kind of bug-fix and the fix is simple without obvious side-effect, so I'm +1 to backport it to 3.9.1.

showuon commented:
> > do you think we should backport this patch to 3.9.1 release?
> 
> This is a kind of bug-fix and the fix is simple without obvious side-effect, so I'm +1 to backport it to 3.9.1.

OK, backported. Thanks.

showuon commented:
Also backported to 4.0 branch.
